### PR TITLE
update widgets in background on startup

### PR DESCRIPTION
**Pull request by djay**
for faster startup show we show even outdated content momentarily in case widgets that use plugins take a long time to load.

### CODE REVIEW

**Comment by drinfernoo** [comment]
While I'm not *against* caching... I think  it is better left to the actual add-on supplying the data. Basically, if an item is cached by AutoWidget, there's a chance (a pretty good one, I'd guess) that it will become outdated "before" the source add-on.

I would especially expect that, considering that users will most often have AutoWidgets on their homescreen (where they will be "used" every time the skin is loaded), and frequently "in lieu of" widgets from a source add-on.

**Comment by djay** [comment]
@drinfernoo just loading 3-4 different plugins on startup (even if they did do caching) takes a lot of time on a low end system like android tv. So I'm waiting a minute to get the homescreen. 
In addition many plugins don't cache since they assume you are only asking for one list and are prepared to wait the 20s it might take to get that data from its downstream api.

**Comment by djay** [comment]
@drinfernoo what if instead of caching, it saved the files each time they were accessed to disk and on kodi startup displayed the saved ones. Then delay the refresh on startup by a few min to allow for teh widgets to be displayed. And when a refresh happens actually access the plugins at this point then update the saved files. This way the skin is never left showing the loading content message for long. 
Perhaps later viewing of a widget might also cause those files to update in the background.
Basically do the updating from the plugins asynchronously in background and the skin always gets the saved version.

**Comment by drinfernoo** [comment]
Now, that seems like an idea. Basically, keep a cache, but only use cached content on first load?

**Comment by djay** [comment]
Yes.
But in addition the subsequent background refresh also updates the cached
files. So the skin should always be reading quickly  from recently updated
cached data rather than waiting.

**Comment by djay** [comment]
@drinfernoo if I do the plan of background caching all the paths and then the skin just loading from the cache. That could be doing a lot more work than needs to be done right? Because the autowidget paths are potentially more than what has been added to the skin? I think skins also don't request the widget to be loaded until its seen so even more updating could be done thats never used. But I think thats a small price to pay for quick loads I think.
An alternative might be that if a path is requested by the skin that is stale, it's displayed but a background refresh is triggered and then a container reload and the skin reloads the newly updated cached paths. But is requesting a skin reload expensive?

**Comment by drinfernoo** [comment]
I think if anything, they should only be loaded (and subsequently cached) whenever the skin loads them.

We want to avoid any unnecessary skin reloads, as any time that happens, **all** the widgets will have to reload, regardless of cached status or source. I've worked pretty hard to fine-tune when and where those refreshes happen... but it's possible we could implement something in the path that indicates when to invalid the cache.

I already do this with most of the paths, by setting the data into a window property, and passing the InfoLabel tag as part of the path. This way, whenever the property gets changed, the *path* changes, triggering Kodi to reload it.

**Comment by djay** [comment]
@drinfernoo that would mean that you might wait 20s for a widget to reload after browsing to it. 
With setting the data into the window property it will only try to reload that single widget?

What if it worked like this.

1. skin requests a path. return cached data no matter how old. save path into a requested queue
2. every 15s process requested queue in the background and update the cache
3. set the window property for each of the updated caches so they reload.

So almost all of the  time the user sees fast loading widgets loaded from disk. if they want the latest they just have to wait a bit for it to reload automatically.

If its not the case you can get the skin to reload a single widget then is it that bad having a complete skin widget reload if eveyrthing is loading from disk? should still be fast right?

**Comment by drinfernoo** [comment]
> If its not the case you can get the skin to reload a single widget then is it that bad having a complete skin widget reload if eveyrthing is loading from disk? should still be fast right?

That's assuming that a user has only AutoWidgets and/or library widgets. If you start mixing in normal add-on widgets, you could quickly bog down a system updating those along with your "fast loading" ones.

**Comment by djay** [comment]
> That's assuming that a user has only AutoWidgets and/or library widgets. If you start mixing in normal add-on widgets, you could quickly bog down a system updating those along with your "fast loading" ones.

Good point. You are correct. 
But can you force a single widget to reload however? Is that what this function does? https://github.com/drinfernoo/plugin.program.autowidget/blob/6d6c257acf2cc97e2e4b5fa7262f35ab88da270e/plugin.program.autowidget/resources/lib/refresh.py#L88

**Comment by drinfernoo** [comment]
Yep, that method updates the window property of whatever value I need to change, triggering Kodi to refresh widgets that have that property in their path.

**Comment by djay** [comment]
@drinfernoo I had a go at implementing it. seems to work.

**Comment by drinfernoo** [comment]
I haven't gotten a chance to test yet, but I like where it's heading. It appears you have gotten around causing a full widget refresh as well, but it also looks like you're updating the cache... always?

**Comment by djay** [comment]
Only the widgets requested by the skin. It puts them into a queue to be updated in the monitor service.  unless it's already been updated in last 5 min (prevent loops). Perhaps could cache it longer.
 If there is no cached version it will update it but not in the background.
I might try switching to using a thread as I don't think this updating
needs to happen in the monitor thread and a local thread will be quicker
and the queue less hacky.
Another seperate issue to consider is how to get lists to update that
chnage based on what was last played? Now that the updating is less
disruptive to the user a simple option would be to just redo all lists
after any playback? Perhaps with some simple algorithm to predict based on
history which lists don't chnage often so less work is done?

**Comment by djay** [comment]
@drinfernoo I think for sure using a local thread is a better idea because the monitor service can load very slowly as I think kodi starts them one by one and lots of plugins have them.

File: plugin.program.autowidget/resources/lib/common/utils.py [hunk 1]
```diff
@@ -392,6 +418,50 @@ def get_files_list(path, titles=None):
...
+        _history = read_json(_history_path)
+        if not _history:
+            _history = {}
+    history = _history.setdefault(hash, [])
+    if add:
+        history.append( (time.time(), add))
+        write_json(_history_path, _history)
+    # predict next update time 
+    return history[-1][0] + 60*5 # just cache until background update is done
+    # if not history:
```
Comment (drinfernoo) [inline plugin.program.autowidget/resources/lib/common/utils.py]: Is this bit down here your (partial?) implementation of the "prediction" algorithm?

File: plugin.program.autowidget/addon.xml [hunk 2]
```diff
@@ -3,6 +3,7 @@
     <requires>
         <import addon="script.module.six" />
         <import addon="script.module.pil" />
+        <import addon="script.module.web-pdb" />
```
Comment (drinfernoo) [inline plugin.program.autowidget/addon.xml]: Before this gets merged, this will need removed, obviously.

File: plugin.program.autowidget/resources/lib/refresh.py [hunk 3]
```diff
@@ -84,6 +88,22 @@ def _refresh(self, startup=False):
             utils.log('+++++ AUTOWIDGET REFRESHING NOT ENABLED +++++',
                       'notice')
 
+def update_cache():
+    while True:
+        widget_id = utils.pop_queue('widgets_to_cache')
+        if not widget_id:
+            break
+        #import web_pdb; web_pdb.set_trace()
```
Comment (drinfernoo) [inline plugin.program.autowidget/resources/lib/refresh.py]: Ditto here.

File: plugin.program.autowidget/resources/lib/common/utils.py [hunk 4]
```diff
@@ -362,18 +377,29 @@ def _get_json_version():
...
-    props_info = info_types + ['customproperties']
-    params = {'jsonrpc': '2.0', 'method': 'Files.GetDirectory',
-              'params': {'properties': info_types if not props else props_info,
-                         'directory': path},
-              'id': 1}
 
-    files = json.loads(call_jsonrpc(json.dumps(params)))
+    # TODO:
+    # - read current cache time and see if its past
+    #import ptvsd; ptvsd.enable_attach(); ptvsd.wait_for_attach()
```
Comment (drinfernoo) [inline plugin.program.autowidget/resources/lib/common/utils.py]: Also these debug lines here.

File: plugin.program.autowidget/resources/lib/common/utils.py [hunk 5]
```diff
@@ -392,6 +418,50 @@ def get_files_list(path, titles=None):
...
+              'id': 1}
+    files_json = call_jsonrpc(json.dumps(params))
+    files = json.loads(files_json)
+    write_json(cache_path, files)
+    expiry = cache_expiry(hash, add=hashlib.sha1(files_json).hexdigest())
+    log("Wrote cache (exp in {}s): {}".format(expiry-time.time(), hash), 'notice')
+    return files
+
+def cache_expiry(hash, add=None):
+    # TODO: clean up no longer used cache files. not read in > 30 days?
```
Comment (drinfernoo) [inline plugin.program.autowidget/resources/lib/common/utils.py]: I would probably use 14 days, as that's the default cache expiry for TMDbHelper (one of the more popular add-ons to use in conjunction with AW).

**Comment by djay** [comment]
@drinfernoo I've now changed to using threads for the background widget updates as it means the widgets get updated faster. I'll fix up the debug code and commented out code then it should be ready to go. I think I'll also store the update history for the caches into seperate files to reduce the risk of race conditions.

**Comment by djay** [comment]
@drinfernoo any chance of updating the project to participate in https://hacktoberfest.digitalocean.com/details?

**Comment by drinfernoo** [comment]
I've added the Hacktoberfest label, which seems to be all for the time being. I should be able to review this PR this week 👌

File: plugin.program.autowidget/resources/lib/common/utils.py [hunk 6]
```diff
@@ -364,37 +377,46 @@ def _get_json_version():
     result = json.loads(call_jsonrpc(json.dumps(params)))['result']['version']
     return (result['major'], result['minor'], result['patch'])
 
-
-def get_files_list(path, titles=None):
-    if not titles:
-        titles = []
+def cache_files(path):
+    hash = hashlib.sha1(path).hexdigest()
```
Comment (drinfernoo) [inline plugin.program.autowidget/resources/lib/common/utils.py:381]: An immediate thing that tripped me up. I'd basically like a way to know which `.cache` file corresponds to which `.widget` file, at the moment. I'd say a tag in the `.widget` file, maybe `cache`, should do just fine.

**Review by drinfernoo — APPROVED** [review]
I've decided to go ahead and merge this. It seems to be working pretty well, and I don't see any glaring issues.

**Comment by drinfernoo** [comment]
I didn't *thoroughly* test this for all the edge cases, or lists that change quite often (as that appears to be what you're planning on attacking in your other PR?), but if there are any issues, I'm sure we'll find them soon... I've merged this and released it as `3.0.0`.

Thanks for your contribution 😁 I think it will make a big difference on loading speeds for many systems, and also increase the value of AutoWidget for widgets of any kind :+1:

**Comment by djay** [comment]
I found some bugs that I've fixed in the other PR. The big one is there seems to be a unicode error where some paths include unicode for some reason so I'll try to get that done ASAP. Sorry I should have let you know friday when I found it to prevent the release.

File: plugin.program.autowidget/resources/lib/common/utils.py (hunk 6 above)
Comment (djay) [reply plugin.program.autowidget/resources/lib/common/utils.py:381]: The cache is actually for a single path or url. A widget can contain many paths (either randomised by autowidget or just paged). In the new PR made a .history file and contains both the history of when the path content changed and which widgets this path was requested by (because in theory you could have multiple widgets with the same content). It's not going to make much sense to put references to the caches in the widget files though.

**Comment by drinfernoo** [comment]
In the future, the best route is to make PRs to `develop`. I have a pretty good group of testers that can easily test `devrepo` releases, which get pushed on every push to `develop`.
